zipline/pipeline/factors/factor.py

```diff
@@ -1229,9 +1236,73 @@ def compute(self, today, assets, out, data):
         # MedianValue.
         median_close10 = MedianValue([USEquityPricing.close], window_length=10)
         median_low15 = MedianValue([USEquityPricing.low], window_length=15)
+
+    A CustomFactor with multiple outputs:
+
+    .. code-block:: python
+
+        class MultipleOutputs(CustomFactor):
+            inputs = [USEquityPricing.close]
+            outputs = ['alpha', 'beta']
+            window_length = N
+
+            def compute(self, today, assets, out, close):
+                computed_alpha, computed_beta = some_function(close)
+                out.alpha = computed_alpha
+                out.beta = computed_beta
+
+        # Each output is returned as its own Factor upon instantiation.
+        alpha, beta = MultipleOutputs()
     '''
     dtype = float64_dtype
 
+    def __getattr__(self, name):
+        if name in self.outputs:
+            return RecarrayFactor(factor=self, attribute=name)
+        else:
+            raise AttributeError(
+                "This factor has no output called '{}'.".format(name)
+            )
+
+    def __iter__(self):
+        if len(self.outputs) < 2:
+            raise ValueError('This factor does not have multiple outputs.')
+        RecarrayFactor_ = partial(RecarrayFactor, self)
+        return iter(map(RecarrayFactor_, self.outputs))
+
+
+class RecarrayFactor(SingleInputMixin, Factor):
+
+    def __new__(cls, factor, attribute):
+        return super(RecarrayFactor, cls).__new__(
+            cls,
+            attribute=attribute,
+            inputs=[factor],
+            outputs=factor.outputs,
+            window_length=0,
+            dtype=factor.dtype,
+            missing_value=factor.missing_value,
+        )
+
+    def _init(self, attribute, *args, **kwargs):
+        self.attribute = attribute
+        return super(RecarrayFactor, self)._init(*args, **kwargs)
+
+    @classmethod
+    def static_identity(cls, attribute, *args, **kwargs):
+        return (
+            super(RecarrayFactor, cls).static_identity(*args, **kwargs),
+            attribute,
+        )
+
+    def _validate(self):
+        super(RecarrayFactor, self)._validate()
+        if not self.outputs:
+            raise TermOutputsNotSpecified(termname=type(self).__name__)
```

Comment:
In thinking about it, I don't think a RecarrayFactor even needs to specify any outputs. At the end of the day it acts how all other regular factors do with a single out. So this can all be removed.